PowerEditor/src/WinControls/ProjectPanel/Directory.cpp

```diff
@@ -0,0 +1,454 @@
+// 3. Links to a library or executes a program that does any of the above.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+#include "Directory.h"
+#include <Shlwapi.h>
+
+Directory::Directory(bool hideEmptyDirs)
+	: _exists(false)
+	, _hideEmptyDirs(hideEmptyDirs)
+	, _wasRead(false)
+{
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	enablePrivileges();
+	_filters.emplace_back(generic_string(TEXT("*.*")));
+}
+
+Directory::Directory(const generic_string& path, const std::vector<generic_string>& filters, bool hideEmptyDirs, bool autoread)
+	: _exists(false)
+	, _path(path)
+	, _filters(filters)
+	, _hideEmptyDirs(hideEmptyDirs)
+	, _wasRead(false)
+{
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	enablePrivileges();
+	if (autoread)
+		read(path);
+}
+
+bool Directory::read(const generic_string& path, const std::vector<generic_string>& filters)
+{
+	_path = path;
+	_exists = false;
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	_files.clear();
+	_dirs.clear();
+	_invisibleDirs.clear();
+
+	if (&_filters != &filters)
+		_filters = filters;
+
+	// first, read directories, never filtered.
+	append(_path, TEXT("*.*"), true);
+
+	// then read files, by each filter
+	if (_filters.empty())
+	{
+		append(_path, TEXT("*.*"), false);
+	}
+	else
+	{
+		for (size_t i=0; i<_filters.size(); ++i)
+			append(_path, _filters[i], false);
+	}
+
+
+	readLastWriteTime(_lastWriteTime);
+
+	_wasRead = true;
+
+	return _exists;
+
+}
+
+bool Directory::readIfChanged(bool respectEmptyDirs)
+{
+	if (!_wasRead)
+	{
+		read();
+		return true;
+	}
+
+	if (respectEmptyDirs && _hideEmptyDirs)
+	{
+		if (!writeTimeHasChanged() && !containsDataChanged())
+			return false;
+	}
+	else
+	{
+		if (!writeTimeHasChanged())
+			return false;
+	}
+
+	const Directory cmp(_path,_filters);
+	if (*this != cmp)
+	{
+		*this = cmp;
+		return true;
+	}
+	return false;
+}
+
+void Directory::append(const generic_string& path, const generic_string& filter, bool readDirs)
+{
+	const bool readFiles = !readDirs;
+
+	generic_string searchPath = _path + TEXT("\\") + filter;
+
+	WIN32_FIND_DATA fd;
+
+	HANDLE hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchNameMatch, NULL, 0);
+
+	if (hFind == INVALID_HANDLE_VALUE)
+		return;
+
+	_exists = true;
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+		const generic_string file(fd.cFileName);
+
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			if (file == TEXT(".") || file == TEXT(".."))
```

Comment:
Here, you could do:

```
if( ( _tcscmp( fd.cFileName, TEXT("..") ) == 0 ) || ( _tcscmp( fd.cFileName, TEXT(".") ) == 0 ) )
```
...so that in the case that we're in an empty directory, we break out of the loop with only two (stack-allocated) string comparisons. If it's a non-empty directory, this is a faster path for the first result returned.